Implement cloud function to support images in multi-class supports [#175363234]

kswenson opened the pull request:
Prior to the introduction of multi-class teacher supports, all CLUE content was scoped to the class in the firebase realtime database, including images. Multi-class teacher supports are stored in firestore and take advantage of firestore's more advanced security rules to enable users in any of a teacher's classes to access the support content. If a multi-class teacher support includes an image stored in the firebase realtime database, however, there's nothing those advanced firestore security rules can do to enable such multi-class access.

Rather than abandoning our image storage mechanism (again!) we implement a google cloud function which reads the multi-class support document from firebase, determines the classes that should have access to the document (and hence to its embedded images) and retrieves the desired image data if and only if the requesting user is in one of the classes that should be allowed access to the image.

Google cloud functions have admin-level access to firebase databases without being subject to any pesky security rules, so they must implement any appropriate security restrictions in code. As such, the `getImageData()` cloud function takes pains to verify that for authenticated users, claims retrieved from the firebase JWT match what is allowed according to the support document stored in firestore. For unauthenticated users (e.g. demo/dev/qa) we verify that what the client passes is appropriate, but since there's no guarantee that a client is following the rules, we at least make sure that unauthenticated users can only access parts of the database that don't require authentication.

Testable at https://collaborative-learning.concord.org/branch/cloud-function/?demo.

@mklewandowski for review, I suggest you focus on the CLUE client side.
@scytacki for review, I suggest you focus on the cloud function itself.

kswenson commented on functions/src/portal-types.ts:
```diff
@@ -0,0 +1,82 @@
+export type UserType = "student" | "teacher";
```
I tried importing these definitions from the original of this file but it confused the build process for the functions, so I ended up making a duplicate of the file here.

scytacki replied:
I think @dougmartin and/or @pjanik ran into something like this before and found a fix.

scytacki commented on functions/src/portal-types.ts:
```diff
@@ -0,0 +1,82 @@
+  offering_id: number;
+}
+
+export interface PortalFirebaseJWTTeacherClaims {
```
Should this extend `PortalFirebaseJWTBaseClaims`?

kswenson replied:
Yes, I believe it should.

scytacki commented on functions/src/index.ts:
```diff
@@ -0,0 +1,128 @@
+  const claims: IDecodedIdToken | undefined = context.auth?.token;
+  const portal: string = claims?.platform_id ||
+                          (_portal === "demo" ? demoName : _portal) ||
+                          "";
```
So my understanding is that an incoming request can make the portal be anything it wants if there aren't claims. But in that case hasValidClaims will be false. So then appMode has to be something other than 'authed'. 

Is it necessary to allow any random appMode, or could that be restricted so anonymous clients can't create new top level keys?

kswenson replied:
I added the appMode restrictions while adding the unit tests.

scytacki reviewed (approved):
This looks good. My only concern is that I'm not seeing any tests of the function. 

The logic seems complex so tests would be good, but I'm OK if this needs to get wrapped up to leave it the way it is.

kswenson commented:
>This looks good. My only concern is that I'm not seeing any tests of the function.

Testing google cloud functions in their entirety (e.g. including database lookups) is fairly involved and requires separate testing projects and/or special credential machinations. That said, the logic you're talking about is more about handling the parameters, so I think perhaps I should look into making that logic testable.

kswenson commented:
>>This looks good. My only concern is that I'm not seeing any tests of the function.

>Testing google cloud functions in their entirety (e.g. including database lookups) is fairly involved and requires separate testing projects and/or special credential machinations. That said, the logic you're talking about is more about handling the parameters, so I think perhaps I should look into making that logic testable.

I refactored the code to use some utility functions and added some unit tests of the utility functions.

mklewandowski commented on functions/src/index.ts:
```diff
@@ -0,0 +1,166 @@
+import { PortalFirebaseJWTClaims } from "./portal-types";
+
+// set to true to enable additional logging
+const DEBUG=false;
```
Spacing around `=`

mklewandowski commented on functions/src/index.ts:
```diff
@@ -0,0 +1,166 @@
+// https://firebase.google.com/docs/functions/typescript
+
+const app = admin.initializeApp({
+  databaseURL: 'https://collaborative-learning-ec215.firebaseio.com/'
```
Double quotes

mklewandowski commented on functions/test/index.rewire.test.ts:
```diff
@@ -0,0 +1,87 @@
+// https://weekly.elfitz.com/2018/10/17/using-rewire-with-typescript-jest/
```
probably can remove one of these comments

mklewandowski commented on src/lib/db.ts:
```diff
@@ -755,7 +761,8 @@ export class DB {
     const { user } = this.stores;
     return new Promise<DBImage>((resolve, reject) => {
-      const imageRef = this.firebase.ref(this.firebase.getImagesPath(user) + "/" + imageKey);
+      const imagePath = this.firebase.getImagesPath(user) + "/" + imageKey;
```
I would have used 
```
...= `${this.firebase.getImagesPath(user)}/${imageKey}`;
```
But, honestly, what you have probably reads better.

mklewandowski reviewed (approved):
👍 LGTM! I don't have much to say here other than a couple of very small code comments.
